[REVIEW] Replace Nulls

aocsa opened the pull request:
This  is a Pull Request. It implements gdf_replace_nulls function.  

gdf_error gdf_replace_nulls(gdf_column* col_out, const gdf_column* new_values_column);

It will take in two gdf_columns.
 * The first one is expected to be a regular gdf_column, the second one has to be a column of the same type as the first, and it has to be of  size one or of the same size as the other column.

 *  case 1: If the second column contains only one value, then this function will replace all nulls in the first column with the value in the second  column.

 * case 2: If the second column is of the same size as the first, then the function will replace all nulls of the first column with the corresponding elements of the  second column 

All the best,
<!--

Thanks for wanting to contribute to cuDF :)

First, if you need some help or want to chat to the core developers, please
visit https://rapids.ai/community.html for links to our Google Group and other
communication channels.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

4. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

5. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_kernel_forwarder {
+  template <typename col_type>
+  void operator()(size_t           nrows,
```
```suggestion
  void operator()(gdf_size_type           nrows,
```

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+struct replace_nulls_kernel_forwarder {
+  template <typename col_type>
+  void operator()(size_t           nrows,
+                  size_t           new_values_length,
```
```suggestion
                  gdf_size_type           new_values_length,
```

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+
+template <typename Type>
+__global__
+void replace_nulls_with_scalar_kernel(int size, Type* vax_data, uint32_t* vax_valid, const Type *vay_data_scalar) 
```
```suggestion
void replace_nulls_with_scalar_kernel(gdf_size_type size, Type* vax_data, uint32_t* vax_valid, const Type *vay_data_scalar) 
```

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+
+template <typename Type>
+__global__
+void replace_nulls_with_column_kernel(int size, Type* vax_data, uint32_t* vax_valid, const Type *vay_data) 
```
```suggestion
void replace_nulls_with_column_kernel(gdf_size_type size, Type* vax_data, uint32_t* vax_valid, const Type *vay_data) 
```

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+    uint32_t is_vax_valid = vax_valid[index];
+
+    uint32_t sel_vax = (is_vax_valid >> position) & 1;
+    vax_data[i] = sel_vax? vax_data[i] : *vay_data_scalar;
```
This should just use `gdf_is_valid` https://github.com/rapidsai/cudf/blob/branch-0.5/cpp/src/utilities/cudf_utils.h#L21

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+    uint32_t is_vax_valid = vax_valid[index];
+
+    uint32_t sel_vax = (is_vax_valid >> position) & 1;
+    vax_data[i] = sel_vax? vax_data[i] : vay_data[i];
```
This should use `gdf_is_valid` to check if a value is valid: https://github.com/rapidsai/cudf/blob/branch-0.5/cpp/src/utilities/cudf_utils.h#L21

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +134,125 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+}
+
+
+//for_each(auto &[data, valid] in zip(col_out->data, col_out->valid) ) {
```
Dead code.

eyalroz commented:
The explanation at the top of this PR seems to be a piece of text copied from some doxygen comment. "This is a binary function"? No, this is a pull request.

Please rephrase it.

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +135,88 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+{
+  GDF_REQUIRE(col_out->dtype == new_values_column->dtype, GDF_DTYPE_MISMATCH);
+  GDF_REQUIRE(new_values_column->size == 1 || new_values_column->size == col_out->size, GDF_COLUMN_SIZE_MISMATCH);
+
```
Needs a few more error checks. Don't want null inputs, and the implementation doesn't support null values in the `new_values_column`.
```suggestion
   GDF_REQUIRE(nullptr != new_values_column->data, GDF_DATASET_EMPTY);
   GDF_REQUIRE(nullptr != col_out->data, GDF_DATASET_EMPTY);
   GDF_REQUIRE(nullptr == new_values_column->valid || 0 == new_values_column->null_count, GDF_VALIDITY_UNSUPPORTED);
```

aocsa replied:
I agree with these changes. I've already added. Thanks

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +135,88 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+
+template <typename Type>
+__global__
+void replace_nulls_with_scalar(gdf_size_type size, Type* out_data, gdf_valid_type * out_valid, const Type *in_data_scalar) 
```
Anonymous namespace.

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +135,88 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+
+template <typename Type>
+__global__
+void replace_nulls_with_column(gdf_size_type size, Type* out_data, gdf_valid_type* out_valid, const Type *in_data) 
```
Anonymous namespace.

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +135,88 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+ *        `replace_nulls` with the apropiate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_kernel_forwarder {
```
Anonymous namespace.

eyalroz commented:
I'm sorry to chime in on this at such a late stage, but I have a few reservations I want to share.

1. Requiring that an input column have only two possible lengths is both confusing and surprising. If there's a scalar variant and a columnar variant, make that clearer/explicit, e.g. with a boolean in the signature or with two separate functions. That's IMHO. Alternatively, consider supporting a fixed-size representation of uniform-value columns, in which case both variants are essentially one and the same.
2. Why is the first argument both an input and an output? Is it not "cleaner" to produce a new column that is equal to the first one on non-nulls and to the replacement otherwise?

I realize the design may already be final, so this is just in case it isn't yet.

jrhemstad commented:
> 1. Requiring that an input column have only two possible lengths is both confusing and surprising. If there's a scalar variant and a columnar variant, make that clearer/explicit, e.g. with a boolean in the signature or with _two separate functions_. That's IMHO. Alternatively, consider supporting a fixed-size representation of uniform-value columns, in which case both variants are essentially one and the same.

Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed). 

>The single-value case specifically can be expressed using the full-column case after applying a fill/replication.

aocsa commented:
> Please rephrase it.

All right

eyalroz commented:
> Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed).

Yes, you did misunderstand. I asked why have the single-value variant when you could fill a column with a single value, then use the full-column variant.

I didn't suggest that the two functions be unified using an extra parameter.

aocsa commented:
> > Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed).
> 
> Yes, you did misunderstand. I asked why have the single-value variant when you could fill a column with a single value, then use the full-column variant.
> 
> I didn't suggest that the two functions be unified using an extra parameter.

In the direction of observation number one of @eyalroz. 

I can change the signature of the `gdf_replace_nulls` adding an extra parameter to specify when the function is used with a scalar and when with a gdf_column.  

In the direction of observation number two of @eyalroz. I agree to change the signature of the  `gdf_replace_nulls`  specifying which one is for input and which one for output. Following the same semantic as `gdf_find_and_replace_all` function. 

Tell me about your thoughts. 
cc @eyalroz @jrhemstad

shwina commented:
FWIW, the scalar case for `find_and_replace` is handled on the Python side:

https://github.com/rapidsai/cudf/blob/branch-0.6/python/cudf/dataframe/series.py#L777-L781

jrhemstad commented:
> > Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed).
> 
> Yes, you did misunderstand. I asked why have the single-value variant when you could fill a column with a single value, then use the full-column variant.
> 
> I didn't suggest that the two functions be unified using an extra parameter.

@eyalroz Ah okay, I did misunderstand. However, that approach would require an unnecessary large allocation + memcpy. 

> In the direction of observation number two of @eyalroz. I agree to change the signature of the `gdf_replace_nulls` specifying which one is for input and which one for output. Following the same semantic as `gdf_find_and_replace_all` function.

@aocsa That's fine with me so long as it still supports an in-place operation, i.e., input and output point to the same `gdf_column`.

> FWIW, the scalar case for `find_and_replace` is handled on the Python side:
@shwina we're moving this to libcudf to be more performant and increase the feature completeness of libcudf.

aocsa commented:
> > > Perhaps I misunderstood your earlier review, but @aocsa's original PR _did_ use two different functions, but you expressed that was unnecessary as it could be expressed in a single function (and I agreed).
> > 
> > 
> > Yes, you did misunderstand. I asked why have the single-value variant when you could fill a column with a single value, then use the full-column variant.

All right! 
> > I didn't suggest that the two functions be unified using an extra parameter.
> 
> @eyalroz Ah okay, I did misunderstand. However, that approach would require an unnecessary large allocation + memcpy.
All right!
> 
> > In the direction of observation number two of @eyalroz. I agree to change the signature of the `gdf_replace_nulls` specifying which one is for input and which one for output. Following the same semantic as `gdf_find_and_replace_all` function.
> 
> @aocsa That's fine with me so long as it still supports an in-place operation, i.e., input and output point to the same `gdf_column`.

I see. So I guess there is not a necessity to add another version. I've just added some basic changes based on the last feedback.  In the last commit, I also rebased the PR, so it is ready to merge. 

All the best.
cc @jrhemstad  

> 
> > FWIW, the scalar case for `find_and_replace` is handled on the Python side:
> > @shwina we're moving this to libcudf to be more performant and increase the feature completeness of libcudf.

jrhemstad commented on cpp/src/replace/replace.cu:
```diff
@@ -134,3 +135,96 @@ gdf_error gdf_find_and_replace_all(gdf_column*       col,
+
+} //end anonymous namespace
+
+gdf_error gdf_replace_nulls(gdf_column* col_out, const gdf_column* new_values_column)
```
`new_values_column` name doesn't agree with name in `functions.h`.

aocsa replied:
done

jrhemstad commented:
rerun tests
